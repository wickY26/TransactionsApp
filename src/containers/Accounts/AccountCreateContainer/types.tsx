import { ReactElement } from 'react';

import { Account } from '../types';

export interface RenderProps {
  onCreateAccount: (data: Account) => void;
}

export interface AccountCreateContainerProps {
  children: (props: RenderProps) => ReactElement;
};
