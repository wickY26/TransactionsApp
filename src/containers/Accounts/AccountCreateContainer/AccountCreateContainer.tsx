import React, { FC, useState } from 'react';
import { Redirect } from 'react-router-dom';
import { useDispatch } from 'react-redux';

import { AccountCreateContainerProps } from './types';
import { thunkCreateAccount } from '../actions';
import { Account } from '../types';

const AccountCreateContainer: FC<AccountCreateContainerProps> = ({ children }) => {
  const [redirect, setRedirect] = useState<boolean>(false);
  const dispatch = useDispatch();

  const handleOnCreateAccount = (account: Account) => {
    dispatch(thunkCreateAccount(account));
    setRedirect(true);
  }

  return (
    <>
      {redirect && <Redirect push to="/accounts" />}
      {children({ onCreateAccount: handleOnCreateAccount })}
    </>
  );
};

export default AccountCreateContainer;
