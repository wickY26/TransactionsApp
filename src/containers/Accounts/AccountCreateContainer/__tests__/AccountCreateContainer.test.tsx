import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import AccountCreateContainer from '..';

const mockAccount = {name:'Poyraz', surname: 'Yilmaz', id: 0};
const mockDispatch = jest.fn();
const mockThunkCreateAccount = jest.fn();

jest.mock('react-redux', () => ({
  useDispatch: () => mockDispatch,
}));

jest.mock('react-router-dom', () => ({
  Redirect: () => 'Redirect',
}));

jest.mock('../../actions', () => ({
  thunkCreateAccount: (acc: any) => mockThunkCreateAccount(acc),
}));

const renderComponent = () => render(
  <AccountCreateContainer>
    {({onCreateAccount}) => (
      <button onClick={()=> onCreateAccount(mockAccount)}>Button</button>
    )}
  </AccountCreateContainer>
);

describe('AccountCreateContainer', () => {

  beforeEach(() => {
    mockThunkCreateAccount.mockReturnValue('CreateUserAction');
  });

  it('should render correctly', () => {
    const component = renderComponent();
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should render redirect when onCreateAccount is called', () => {
    const component = renderComponent();
    fireEvent.click(screen.getByText('Button'));
    expect(component.asFragment()).toMatchSnapshot();
  });
  
  it('should dispatch createAccountAction when onCreateAccount is called', () => {
    renderComponent();
    fireEvent.click(screen.getByText('Button'));
    expect(mockThunkCreateAccount).toHaveBeenCalledWith(mockAccount);
    expect(mockDispatch).toHaveBeenCalledWith('CreateUserAction');
  });

});
