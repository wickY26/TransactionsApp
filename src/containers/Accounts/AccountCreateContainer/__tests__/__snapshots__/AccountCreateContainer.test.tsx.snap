// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountCreateContainer should render correctly 1`] = `
<DocumentFragment>
  <button>
    Button
  </button>
</DocumentFragment>
`;

exports[`AccountCreateContainer should render redirect when onCreateAccount is called 1`] = `
<DocumentFragment>
  Redirect
  <button>
    Button
  </button>
</DocumentFragment>
`;
